dot/state/epoch.go

```diff
@@ -357,28 +326,41 @@ func (s *EpochState) setLatestConfigData(epoch uint64) error {
 	return s.db.Put(latestConfigDataKey, buf)
 }
 
-// GetConfigData returns the config data for a given epoch persisted in database
-// otherwise tries to get the data from the in-memory map using the header.
-// If the header params is nil then it will search only in the database
-func (s *EpochState) GetConfigData(epoch uint64, header *types.Header) (*types.ConfigData, error) {
-	configData, err := s.getConfigDataFromDatabase(epoch)
-	if err == nil && configData != nil {
-		return configData, nil
+// GetConfigData returns the newest config data for a given epoch persisted in database
+// otherwise tries to get the data from the in-memory map using the header. If we don't
+// find any config data for the current epoch we lookup in the previous epochs, as the spec says:
+// - The supplied configuration data are intended to be used from the next epoch onwards.
+// If the header params is nil then it will search only in the database.
+func (s *EpochState) GetConfigData(epoch uint64, header *types.Header) (configData *types.ConfigData, err error) {
+	if header != nil {
+		s.nextConfigDataLock.RLock()
+		defer s.nextConfigDataLock.RUnlock()
 	}
 
-	if err != nil && !errors.Is(err, chaindb.ErrKeyNotFound) {
-		return nil, fmt.Errorf("failed to get config data from database: %w", err)
-	} else if header == nil {
-		// if no header is given then skip the lookup in-memory
-		return configData, nil
-	}
+	for tryEpoch := epoch; tryEpoch >= 0; tryEpoch-- {
+		configData, err = s.getConfigDataFromDatabase(tryEpoch)
+		if err == nil && configData != nil {
+			return configData, nil
```

Comment:
Are you sure about this part? I seems like we are returning config data for a different epoch that the one asked in the argument. 
I did not find anything explicit in the spec

Reply:
In the `polkadot-host-spec_v0.1.1.pdf` in the section `6.1.2 Runtime-to-Consensus Engine Message` there is the description of the `Next Config Data` which says:

> - Next Config Data: These messages are only issued on configuration change and in the first block of an epoch. The supplied configuration data are intended to be used from the next epoch onwards.

The next config data is not issued every epoch as next epoch data. So in the spec when they say:  _The supplied configuration data are intended to be used from the next epoch onwards_ means we should use the closest config data for epoch `X`, if epoch `X` does not contain any config data we should lookup in the epoch `X-1` trying to retrieve the config data.

Reply:
thanks!